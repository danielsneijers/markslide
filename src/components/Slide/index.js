// @flow
import React, { PureComponent } from 'react'
import { debounce } from 'lodash'
import classNames from 'classnames'
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup'
import { elementFitsViewport, scaleElementToFit } from 'utils/viewport'
import CSS from './style.css'

type Props = {
  index: number,
  content: string,
  className?: string
}

type State = {
  style: {
    transform?: string,
    position?: string,
    width?: string
  }
}

const SLIDE_ID: string = 'slide-content'

class Slide extends PureComponent {
  props: Props
  state: State

  initialState = { style: {} }
  state = this.initialState

  componentDidMount () {
    this.checkForScaling()

    window.addEventListener('resize', this.checkForScaling)
  }

  componentWillReceiveProps () {
    this.setState(this.initialState)
    this.checkForScaling()
  }

  componentWillUnMount () {
    this.checkForScaling.cancel()
    window.removeEventListener('resize', this.checkForScaling)
  }

  checkForScaling: Function = debounce(() => {
    const wrapper = document.getElementById(SLIDE_ID)
    const fitsViewport = elementFitsViewport(wrapper)
    const scaleFraction = fitsViewport ? 1 : scaleElementToFit(wrapper)
    const position = fitsViewport ? 'relative' : 'absolute'
    const width = fitsViewport ? 'auto' : 90 / scaleFraction + '%'

    const style = {
      transform: `scale(${scaleFraction})`,
      opacity: 1,
      position,
      width
    }

    this.setState({ style })
  }, 100)

  render () {
    const { content, index, className } = this.props
    const slideClasses = classNames(
      CSS.slide,
      className,
      'regular-slide',
      `slide-${index}`
    )

    return (
      <CSSTransitionGroup
        transitionName='slide'
        transitionAppearTimeout={500}
        transitionEnterTimeout={500}
        transitionLeaveTimeout={300}
        transitionAppear
        component='div'
        className={slideClasses}
        key={`slide-${index}`}
      >
        <div
          id={SLIDE_ID}
          style={{ opacity: 0, ...this.state.style }}
          dangerouslySetInnerHTML={{ __html: content }}
        />
      </CSSTransitionGroup>
    )
  }
}

export default Slide
