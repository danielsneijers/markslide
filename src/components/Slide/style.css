.slide {
  height: 100vh;
  width: 100vw;
  background-color: #fafafa;
  padding: 10%;
  display: flex;
  align-items: center;
  justify-content: center;
}

:global .slide-appear {
  opacity: 0.01;
}

:global .slide-appear.slide-appear-active {
  opacity: 1;
  transition: opacity .5s ease-in;
}

:global .slide-enter {
  opacity: 0.01;
}

:global .slide-enter.slide-enter-active {
  opacity: 1;
  transition: opacity 500ms ease-in;
}

:global .slide-leave {
  opacity: 1;
}

:global .slide-leave.slide-leave-active {
  opacity: 0.01;
  transition: opacity 300ms ease-in;
}
