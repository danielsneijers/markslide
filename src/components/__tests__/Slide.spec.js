import React from 'react'
import { shallow } from 'enzyme'
import Slide from 'components/Slide'

describe('components/Slide', () => {
  it('renders correctly', () => {
    const content = <h1>Hello Test!</h1>
    const tree = shallow(<Slide content={content} index={1} />)

    expect(tree).toMatchSnapshot()
  })
})
