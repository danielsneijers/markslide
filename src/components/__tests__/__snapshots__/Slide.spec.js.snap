// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Slide renders correctly 1`] = `
<CSSTransitionGroup
  component="div"
  transitionAppear={true}
  transitionAppearTimeout={500}
  transitionEnter={true}
  transitionEnterTimeout={500}
  transitionLeave={true}
  transitionLeaveTimeout={300}
  transitionName="slide"
>
  <div
    className="slide regular-slide slide1"
    dangerouslySetInnerHTML={
      Object {
        "__html": <h1>
          Hello Test!
      </h1>,
      }
    }
  />
</CSSTransitionGroup>
`;
